source/2_platform/rex_directx/src/directx_renderer.cpp

```diff
@@ -73,76 +238,1816 @@ namespace rex
+        }
+
+        // Create depth/stencil view to bind to the pipeline and
+        // depth/stencil buffer as resource to write depth values to
+        if(internal::create_dsv_for_swapchain(userData.window_width, userData.window_height, depthTargetSlot) == false)
+        {
+          REX_ERROR(LogDirectX, "Failed to create dsv");
+          return false;
+        }
+
+        // Execute the swapchain buffer creation commands.
+        if(internal::close_command_list() == false)
+        {
+          REX_ERROR(LogDirectX, "Failed to close command list");
+          return false;
+        }
+
+        internal::exec_command_list();
+
+        // Wait until creation is complete.
+        flush_command_queue();
+
+        // Create viewport to render our image in
+        // A viewport always needs to reset whenever a command list is reset
+        g_ctx.screen_viewport.TopLeftX = 0.0f;
+        g_ctx.screen_viewport.TopLeftY = 0.0f;
+        g_ctx.screen_viewport.Width    = static_cast<f32>(userData.window_width);
+        g_ctx.screen_viewport.Height   = static_cast<f32>(userData.window_height);
+        g_ctx.screen_viewport.MinDepth = 0.0f;
+        g_ctx.screen_viewport.MaxDepth = 1.0f;
+
+        // Cull pixels drawn outside of the backbuffer ( such as UI elements )
+        g_ctx.scissor_rect = {0, 0, static_cast<s32>(userData.window_width), static_cast<s32>(userData.window_height)};
+
+        directx::set_debug_name_for(g_ctx.command_allocator.Get(), "Global Command Allocator");
+        directx::set_debug_name_for(g_ctx.command_list.Get(), "Global Command List");
+        directx::set_debug_name_for(g_ctx.command_queue.Get(), "Global Command Queue");
+        directx::set_debug_name_for(g_ctx.fence.Get(), "Global Fence");
+
+        return true;
+      }
+
+      //-------------------------------------------------------------------------
+      void shutdown()
+      {
+        if(g_ctx.device != nullptr)
+        {
+#if REX_DEBUG
+            bool can_report_dxgi_live_objects = false;
+            bool can_report_dx12_live_objects = false;
+
+            // DXGI
+            wrl::com_ptr<IDXGIDebug1> dxgi_debug;
+            if (FAILED(DXGIGetDebugInterface1(0, IID_PPV_ARGS(dxgi_debug.GetAddressOf()))))
+            {
+                REX_WARN(LogDirectX, "Unable to Query DXGI Debug Interface");
+            }
+            else
+            {
+                can_report_dx12_live_objects = true;
+            }
+
+            // DX12
+            wrl::com_ptr<ID3D12DebugDevice> dx12_debug;
+            if (FAILED(g_ctx.device->QueryInterface(IID_PPV_ARGS(&dx12_debug))))
+            {
+                REX_WARN(LogDirectX, "Unable to Query DX12 Debug Interface");
+            }
+            else
+            {
+                can_report_dxgi_live_objects = true;
+            }
+#endif
+
+            internal::release_frame_resources(g_ctx.frame_ctx.frame_resources.data(), g_ctx.frame_ctx.frame_resources.size());
```

Comment:
Once I convert them into pointers instead of shared copies we do not no, but now this hold a reference to back buffer render targets. if we don't release these manually they will get cleared after main, when all allocated memory has already been destroyed, this will result in a crash

Reply:
linked to https://github.com/New-Creta/Rex/issues/268